### PR TITLE
Nico Shanstrom

### PR DESCRIPTION
**Pull request by NicoShanstrom**
I am having problems figuring out when and how to create the join tables and what they should have on them. I know I know this, but for some reason I get confused every time as to what i need to create and then the relationships with that. I was like just throwing up all the relationships i could, and then had to add another later on in the stories. Other than that i would have had a much better time going through the stories, but it took me over an hour to get off the ground and then stumbled on some relationships later bc my tables were messed up. Also, i think i was a little confused as to what exactly the end of user story 3 was asking. I just created another mechanic with the same name as another and then made a method to pluck unique names. I was not sure how to go about creating the object for this one.

### CODE REVIEW

File: app/controllers/amusement_parks_controller.rb [hunk 1]
```diff
@@ -0,0 +1,5 @@
+class AmusementParksController < ApplicationController
+  def show
+    @amusement_park = AmusementPark.find(params[:id])
+  end
+end
```
Comment (abdulredd) [inline app/controllers/amusement_parks_controller.rb:5]: 👍

File: app/controllers/mechanics_controller.rb [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+    @mechanic = Mechanic.find(params[:id])
+  end
+
+  def update
+    # require 'pry'; binding.pry
+    @mechanic = Mechanic.find(params[:id])
+    @ride = Ride.find(params[:mechanic][:ride_id])
+    @mechanic.rides << @ride
+    redirect_to mechanic_path(@mechanic)
+  end
```
Comment (abdulredd) [inline app/controllers/mechanics_controller.rb:12]: When looking at US2, here is what we're being asked to do: `"When I go to a mechanic's show page, I see a form to ADD a ride to their workload with an ID of an existing ride"`.

I can see that you're trying to do that using an `update` but because we are CREATING a new relationship we want this to in the `create` action.

The only other thing is where this operation needs to take place.  Since `a mechanic can HAVE MANY rides` and `a ride can HAVE MANY mechanics`, when adding an additional ride to a mechanic, this operation should happen in our joins controller, `MechanicRidesController`.

Here is what the code could look like:

```ruby
class MechanicRidesController < ApplicationController
  def create
    @mechanic = Mechanic.find(params[:id])
    @ride = Ride.find(params[:mechanic][:ride_id])
    @mechanic.rides << @ride
    redirect_to mechanic_path(@mechanic)
  end
end
```

So long story short, the implementation is correction, the location of both the action and controller aren't.

File: app/models/amusement_park.rb [hunk 3]
```diff
@@ -1,3 +1,8 @@
 class AmusementPark < ApplicationRecord
   has_many :rides
+  has_many :mechanics
```
Comment (abdulredd) [inline app/models/amusement_park.rb:3]: An amusement park does not have a direct connection to a mechanic and the only way to get there is `through` `:rides`.

```ruby
has_many :rides
has_many :mechanics, through: :rides
```

This is one reason why I recommend creating a dbdiagram.io schema so you can actually see the table connections which might make the associations easier to see.

File: app/models/amusement_park.rb [hunk 4]
```diff
@@ -1,3 +1,8 @@
...
+
+  def unique_name
+    mechanics.distinct.pluck(:name)
+  end
```
Comment (abdulredd) [inline app/models/amusement_park.rb:7]: I cannot imagine that this method actually works. I believe you should be getting an error saying that the relationship between an amusement park and a mechanic does not exist. This is unless you added that relationship yourself on the `mechanics` and `amusement_parks` tables.

In any case, the AR is correct if you got the `mechanics` connected properly.

File: app/models/mechanic.rb [hunk 5]
```diff
@@ -0,0 +1,6 @@
+class Mechanic < ApplicationRecord
+  belongs_to :amusement_park
+  has_many :mechanic_rides
+  has_many :rides
+  has_many :rides, through: :mechanic_rides
```
Comment (abdulredd) [inline app/models/mechanic.rb:5]: Ok so it does look like you added a connection between a mechanic and an amusement park. This unfortunately is incorrect and not necessary. Let's take a look at the setup instructions for the db.

These 2 entities have a one-to-many relationship with each other:
```
* AmusementParks have many Rides (has_many :rides)
* Rides belong to an AmusementPark (belongs_to :amusement_park)
```

These 2 entities have a many-to-many relationship with each other so you need a joins table between them. (ie. `MechanicRides`.
```
* Mechanics can work on many Rides (has_many :mechanic_rides / has_many :rides, through: :mechanic_rides)
* Rides can have many Mechanics working on them (has_many :mechanic_rides / has_many :mechanics, through: :mechanic_rides)
```

Any other extra associations would need to be created with a `through` relationship.

The most important part of any IC or project is going to be the database setup and if you get stuck on that then I ask that you reach out.

File: app/models/mechanic_ride.rb [hunk 6]
```diff
@@ -0,0 +1,4 @@
+class MechanicRide < ApplicationRecord
+  belongs_to :mechanic
+  belongs_to :ride
+end
```
Comment (abdulredd) [inline app/models/mechanic_ride.rb:4]: 👍

File: app/models/ride.rb [hunk 7]
```diff
@@ -1,3 +1,5 @@
 class Ride < ApplicationRecord
   belongs_to :amusement_park
+  has_many :mechanic_rides
+  has_many :mechanics, through: :mechanic_rides
```
Comment (abdulredd) [inline app/models/ride.rb:4]: 👍

File: app/views/amusement_parks/show.html.erb [hunk 8]
```diff
@@ -0,0 +1,13 @@
...
+<h2>Park Mechanics</h2>
+<section class='mechanics'>
+  <ol>
+    <% @amusement_park.unique_name.each do |mechanic| %>
+    <li>
+      <h3>Mechanic name: <%= mechanic %></h3>
+    </li>
+    <%end%>
+  </ol>
+</section>
```
Comment (abdulredd) [inline app/views/amusement_parks/show.html.erb:13]: In your block variable I would make it more descriptive, it looks like you're pulling the entire `mechanic` object just looking at the name. I would call it `mechanic_name` or something like that to make it obvious in your code. I was ready to say that you weren't pulling out the name on line 9 with `mechanic.name` before having to go back to the model and see that you were using `pluck(:name)` which is fine but make the block variable representative of that.

File: app/views/mechanics/show.html.erb [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+
+<h2>Add existing ride to <%=@mechanic.name%> workload</h2>
+  <section class='addride'>
+    <%=form_with model: [@ride, @mechanic], local: true do |form|%>
+    <%= form.label :ride_id, "Ride Id:" %>
+    <%= form.text_field :ride_id %>
+
+    <%= form.submit "Add ride" %>
+    <%end%>
+  </section>
```
Comment (abdulredd) [inline app/views/mechanics/show.html.erb:23]: The only issue I see with this view is the path for the form. I think you have the variables backward. `@mechanic` should come before `@ride` since we're creating a ride for a mechanic.

My only other question is where the `@ride` variable is coming from since you did not set it up in the `show` action in the `MechanicsController` so I would imagine this value is `nil`. I would stick with simple `form_with model:` that deals with just 1 model for now and for anything else use `form_with url:` until you get more comfortable with the `form_with model:` syntax and passing multiple objects.

File: config/routes.rb [hunk 10]
```diff
@@ -3,4 +3,6 @@
 
   # Defines the root path route ("/")
   # root "articles#index"
+  resources :mechanics, only: [:show, :update]
+  resources :amusement_parks, only: [:show]
```
Comment (abdulredd) [inline config/routes.rb:7]: SInce we are not updating a specific mechanic and instead creating a new relationship between a mechanic and a ride, you would not have a `:update` route for a mechanic here. Here is what your routes should look like:

```ruby
resources :mechanics, only: [:show] do
  resources :rides, only: [:create], controller: :mechanic_rides
end

resources :amusement_parks, only: [:show]
```

File: spec/features/amusement_parks/show_spec.rb [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+      # Then I see the name and price of admissions for that amusement park
+      expect(page).to have_content(@park1.name)
+      expect(page).to have_content("Admission cost: #{@park1.admission_cost}")
+      # And I see the names of all mechanics that are working on that park's rides,
+      # And I see that the list of mechanics is unique
+      within '.mechanics' do
+        expect(page).to have_content(@mech1.name, count: 1)
+        expect(page).to have_content(@mech2.name, count: 1)
+      end
+    end
```
Comment (abdulredd) [inline spec/features/amusement_parks/show_spec.rb:34]: Testing has always been one of your strengths and continues to be, great job using "anchor" text for the admission cost and using `within` to target a specific part of the page. I also love how you used `count: 1` to verify uniqueness.

File: spec/features/mechanics/show_spec.rb [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+      end
+      # I’m taken back to that mechanic's show page
+      expect(current_path).to eq(mechanic_path(@mech1))
+      # And I see the name of that newly added ride on this mechanic's show page.
+      within '.rides' do
+        expect(page).to have_content("Ride name: #{@ride1.name}")
+        expect(page).to have_content("Ride name: #{@ride2.name}")
+        expect(page).to have_content("Ride name: #{@ride3.name}")
+      end
+    end
```
Comment (abdulredd) [inline spec/features/mechanics/show_spec.rb:48]: While this test covers the majority of the US, I think it could have benefited from first checking that the name of the ride isn't already on the page with a `to_not have_content` assertion before filling out the form.

File: spec/models/amusement_park_spec.rb [hunk 13]
```diff
@@ -1,7 +1,32 @@
...
+  
+      @mech1_ride1 = MechanicRide.create!(mechanic_id: @mech1.id, ride_id: @ride1.id)
+      @mech1_ride2 = MechanicRide.create!(mechanic_id: @mech1.id, ride_id: @ride2.id)
+      @mech2_ride1 = MechanicRide.create!(mechanic_id: @mech2.id, ride_id: @ride1.id)
+      @mech2_ride2 = MechanicRide.create!(mechanic_id: @mech2.id, ride_id: @ride3.id)
+  end
+
   describe "relationships" do
     it { should have_many(:rides) }
   end
```
Comment (abdulredd) [inline spec/models/amusement_park_spec.rb:23]: You added a new association to the `AmusementPark` model but didn't test it.

File: spec/models/mechanic_spec.rb [hunk 14]
```diff
@@ -0,0 +1,7 @@
+require 'rails_helper'
+
+RSpec.describe Mechanic, type: :model do
+  describe "relationships" do
+    it { should have_many(:rides).through(:mechanic_rides) }
+  end
```
Comment (abdulredd) [inline spec/models/mechanic_spec.rb:6]: There were quite a few associations that got added to the `Mechanic` model that went untested.

File: spec/models/ride_spec.rb [hunk 15]
```diff
@@ -3,5 +3,6 @@
 RSpec.describe Ride, type: :model do
   describe "relationships" do
     it { should belong_to(:amusement_park) }
+    it { should have_many(:mechanics) }
   end
```
Comment (abdulredd) [inline spec/models/ride_spec.rb:7]: Missing association tests.

**Review by abdulredd — COMMENTED** [review]
Hey Nico, it looks like what derailed a lot of your progress was the database setup and just looking at your migrations and your models it looks like you went back and forth as to what was needed and what wasn't and made the schema a little more complex than it needed to be. You also added a lot of extra associations and then did not test those which dropped your testing to below a meets expectations.


In the future, if you are having issues with the setup, please reach out to the instructors so we can help guide you in the right direction.
